### ISSUE TITLE
SSHClientConnection not closing listeners on error when using forward_local_port()

### ISSUE THREAD

**Issue by rmawatson** [ronf/asyncssh#309]
I am using the following example from the documentation

```
import asyncio, asyncssh, sys

async def run_client():
    async with asyncssh.connect('my.host.com') as conn:
        listener = await conn.forward_local_port('', 9292, 'other.host.com', 10040)
        await listener.wait_closed()
try:
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))
```

When I kill the ssh connection on the remote server with `ss -K ..` the logs report the connection as closed, but I get no exceptions raised and wait_closed() does not return.

```
...
DEBUG:asyncssh:[conn=0] Got keepalive response
DEBUG:asyncssh:[conn=0] Sending keepalive request
DEBUG:asyncssh:[conn=0] Got keepalive response
INFO:asyncssh:[conn=0] Connection failure: [WinError 64] The specified network name is no longer available
```
*Edit*

There doesn't seem to be any way to pass in an error handler function, but manually adding one, and using this to close the listener on error does fix the problem. 

```
    async with asyncssh.connect('my.host.com') as conn:

        listener = await conn.forward_local_port('', 9292, 'other.host.com', 10040)

        def close_listener(*args,**kwargs):
            print("close listener")
            listener.close()
        conn._error_handler = close_listener

        await listener.wait_closed()
```
It does seem like the listener should be automatically closed when the connection is closed, but perhaps not, and the current behaviour is as intended?

*Edit*

It seems the proper way to get notifications on connection status is use `SSHClientConnectionOptions(client_factory=...)`, sublcassing `SSHClient` overriding `connection_made` storing the `SSHClientConnection` connection instance that is passed and overriding `connection_lost` to handle the error.

However after playing around it seems like the listener is not added to the `self._local_listeners` dictionary on the `SSHClientConnection` when using `forward_local_port()`. The dictionary is iterated in `_cleanup` [here](https://github.com/ronf/asyncssh/blob/master/asyncssh/connection.py#L575-L576), where the listeners are all closed.

adding `self._local_listeners[listen_host, listen_port] = listener` [here](https://github.com/ronf/asyncssh/blob/master/asyncssh/connection.py#L2436) solves the problem and it now exits properly on connection failure.

The same problem possibly exists in `forward_local_path` although I have not tested it.

Fixed by [#310](https://github.com/ronf/asyncssh/pull/310)

**Comment by ronf** [comment]
Thank you for the report!

I was able to reproduce the issue here, and I agree with your proposed solution of adding the local listener to self._local_listeners. That's already happening in the case of an SSHServerConnection being asked to open a listener when processing a 'tcpip-forward' global request, and it is cleaned up either when the listener is canceled, or automatically in _cleanup when the connection breaks, if any listeners are still open at that point. Doing something similar on the client side would make sense, but I need to look into how to make sure listeners are removed from _local_listeners when they are closed explicitly by the application.

In addition to the local listener opened in forward_local_port(), I think a similar treatment is needed for forward_local_path() as you said, and I also need to look into whether something similar should be done in forward_socks(), create_agent_listener() called when SSH agent forwarding is requested, and possible in attach_x11_listener() which is called to set up X11 forwarding.

**Comment by ronf** [comment]
After a bit of code examination, it looks like create_agent_listener() should be fine as-is. That has its own _agent_listener member in SSHServerConnection and it already gets closed automatically by _cleanup(). For the X11 listener, I think we might be ok as well, as it should automatically close its _tcp_listener when all of the channels which requested it have been cleaned up, and a connection close should trigger _cleanup() to be called on all open channels.

As for forward_socks(), it follows the pattern of forward_local_port() and forward_local_path(), returning the listener to the caller but not keeping track in the connection object of any such listeners being opened. So, I think all three of these functions probably need this fix.

**Comment by ronf** [comment]
Ok - here's a potential fix:

```patch
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -2433,15 +2433,17 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
                                                          tunnel_connection,
                                                          listen_host,
                                                          listen_port)
-
-            if dest_port == 0:
-                dest_port = listener.get_port()
-
-            return listener
         except OSError as exc:
             self.logger.debug1('Failed to create local TCP listener: %s', exc)
             raise
 
+        if listen_port == 0:
+            listen_port = listener.get_port()
+
+        self._local_listeners[listen_host, listen_port] = listener
+
+        return listener
+
     @async_context_manager
     async def forward_local_path(self, listen_path, dest_path):
         """Set up local UNIX domain socket forwarding
@@ -2474,13 +2476,25 @@ class SSHConnection(SSHPacketHandler, asyncio.Protocol):
                          listen_path, dest_path)
 
         try:
-            return await create_unix_forward_listener(self, self._loop,
-                                                      tunnel_connection,
-                                                      listen_path)
+            listener = await create_unix_forward_listener(self, self._loop,
+                                                          tunnel_connection,
+                                                          listen_path)
         except OSError as exc:
             self.logger.debug1('Failed to create local UNIX listener: %s', exc)
             raise
 
+        self._local_listeners[listen_path] = listener
+
+        return listener
+
+    def close_forward_listener(self, listen_key):
+        """Mark a local forwarding listener as closed"""
+
+        listener = self._local_listeners.get(listen_key)
+
+        if listener:
+            del self._local_listeners[listen_key]
+
 
 class SSHClientConnection(SSHConnection):
     """SSH client connection
@@ -4018,12 +4032,20 @@ class SSHClientConnection(SSHConnection):
                          (listen_host, listen_port))
 
         try:
-            return await create_socks_listener(self, self._loop, tunnel_socks,
-                                               listen_host, listen_port)
+            listener = await create_socks_listener(self, self._loop,
+                                                   tunnel_socks,
+                                                   listen_host, listen_port)
         except OSError as exc:
             self.logger.debug1('Failed to create local SOCKS listener: %s', exc)
             raise
 
+        if listen_port == 0:
+            listen_port = listener.get_port()
+
+        self._local_listeners[listen_host, listen_port] = listener
+
+        return listener
+
     @async_context_manager
     async def start_sftp_client(self, env=(), send_env=(),
                                 path_encoding='utf-8', path_errors='strict'):
diff --git a/asyncssh/listener.py b/asyncssh/listener.py
index 23d741e..46fc959 100644
--- a/asyncssh/listener.py
+++ b/asyncssh/listener.py
@@ -191,10 +191,12 @@ class SSHUNIXClientListener(SSHClientListener):
 class SSHForwardListener(SSHListener):
     """A listener used when forwarding traffic from local ports"""
 
-    def __init__(self, servers, listen_port=0):
+    def __init__(self, conn, servers, listen_key, listen_port=0):
         super().__init__()
 
+        self._conn = conn
         self._servers = servers
+        self._listen_key = listen_key
         self._listen_port = listen_port
 
     def get_port(self):
@@ -205,8 +207,13 @@ class SSHForwardListener(SSHListener):
     def close(self):
         """Close this listener"""
 
-        for server in self._servers:
-            server.close()
+        if self._conn:
+            self._conn.close_forward_listener(self._listen_key)
+
+            for server in self._servers:
+                server.close()
+
+            self._conn = None
 
     async def wait_closed(self):
         """Wait for this listener to finish closing"""
@@ -274,7 +281,8 @@ async def create_tcp_local_listener(conn, loop, protocol_factory,
         server = await loop.create_server(protocol_factory, sock=sock)
         servers.append(server)
 
-    return SSHForwardListener(servers, listen_port)
+    listen_key = listen_host, listen_port
+    return SSHForwardListener(conn, servers, listen_key, listen_port)
 
 
 async def create_tcp_forward_listener(conn, loop, coro,
@@ -300,7 +308,7 @@ async def create_unix_forward_listener(conn, loop, coro, listen_path):
 
     server = await loop.create_unix_server(protocol_factory, listen_path)
 
-    return SSHForwardListener([server])
+    return SSHForwardListener(conn, [server], listen_path)
 
 
 async def create_socks_listener(conn, loop, coro, listen_host, listen_port):
```

I'm thinking about adding some new unit tests here before I check this in, though. Also, I need to track down a traceback in the existing unit tests about failing to close transports that I think might be a sign of something wrong even though the existing tests are still passing.

**Comment by rmawatson** [comment]
Thanks for taking a look ronf. 

Applied your more complete patch today and all seems fine for us so far. Thank you.

Would be great to see this rolled out into a 2.4.1 bump on PyPI.

**Comment by ronf** [comment]
Thanks for letting me know!

I'm still working on an unidentified issue showing up in the unit tests, but once I resolve that this should be good to go, and if nothing else pressing comes up in the meantime I can probably cut a minor release for this.

**Comment by ronf** [comment]
Ok - the final version of this fix is now available in the "develop" branch as commit 869cea0 if you want to pick up the code from there.

If everything looks good on the CI builds, I'll look into cutting a new release.

**Comment by ronf** [comment]
This fix is now available in AsyncSSH 2.4.1. Thanks again for your help here!

**Comment by rmawatson** [comment]
Thanks for this - working great.
